File: app/src/main/java/fr/insapp/insapp/adapters/EventRecyclerViewAdapter.java
package fr.insapp.insapp.adapters;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;
import fr.insapp.insapp.R;
import fr.insapp.insapp.activities.ClubActivity;
import fr.insapp.insapp.http.ServiceGenerator;
import fr.insapp.insapp.models.Club;
import fr.insapp.insapp.models.Event;
import fr.insapp.insapp.models.EventComparator;
import jp.wasabeef.glide.transformations.RoundedCornersTransformation;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by thomas on 18/11/2016.
 */

public class EventRecyclerViewAdapter extends BaseRecyclerViewAdapter<EventRecyclerViewAdapter.EventViewHolder> {

    protected List<Event> events;

    private int layout;

    private OnEventItemClickListener listener;

    public interface OnEventItemClickListener {
        void onEventItemClick(Event event);
    }

    public EventRecyclerViewAdapter(Context context, int layout) {
        this.context = context;
        this.events = new ArrayList<>();
        this.layout = layout;
    }

    public void setOnItemClickListener(OnEventItemClickListener listener) {
        this.listener = listener;
    }

    public void addItem(Event event) {
        this.events.add(event);
        Collections.sort(events, new EventComparator());

        this.notifyDataSetChanged();
    }

    public void updatePost(int id, Event event){
        this.events.set(id, event);
        this.notifyItemChanged(id);
    }

    @Override
    public EventRecyclerViewAdapter.EventViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext()).inflate(layout, parent, false);
        return new EventRecyclerViewAdapter.EventViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final EventRecyclerViewAdapter.EventViewHolder holder, int position) {
        final Event event = events.get(position);

        if (layout == R.layout.row_event_with_avatars) {
            Call<Club> call = ServiceGenerator.create().getClubFromId(event.getAssociation());
            call.enqueue(new Callback<Club>() {
                @Override
                public void onResponse(@NonNull Call<Club> call, @NonNull Response<Club> response) {
                    if (response.isSuccessful()) {
                        final Club club = response.body();

                        Glide
                                .with(context)
                                .load(ServiceGenerator.CDN_URL + club.getProfilePicture())
                                .crossFade()
                                .into(holder.avatar);

                        holder.avatar.setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View view) {
                                context.startActivity(new Intent(context, ClubActivity.class).putExtra("club", club));
                            }
                        });
                    }
                    else {
                        Toast.makeText(context, "EventRecyclerViewAdapter", Toast.LENGTH_LONG).show();
                    }
                }

                @Override
                public void onFailure(@NonNull Call<Club> call, @NonNull Throwable t) {
                    Toast.makeText(context, "EventRecyclerViewAdapter", Toast.LENGTH_LONG).show();
                }
            });
        }

        Glide.with(context)
                .load(ServiceGenerator.CDN_URL + event.getImage())
                .bitmapTransform(new CenterCrop(context), new RoundedCornersTransformation(context, 8, 0))
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(holder.thumbnail);

        holder.name.setText(event.getName());

        final int nb_participants = (event.getAttendees() == null) ? 0 : event.getAttendees().size();
        if (nb_participants <= 1) {
            holder.participants.setText(Integer.toString(nb_participants) + " participant");
        }
        else {
            holder.participants.setText(Integer.toString(nb_participants) + " participants");
        }

        final int diffInDays = (int) ((event.getDateEnd().getTime() - event.getDateStart().getTime()) / (1000 * 60 * 60 * 24));
        if (diffInDays < 1 && event.getDateStart().getMonth() == event.getDateEnd().getMonth()) {
            DateFormat dateFormat_oneday = new SimpleDateFormat("'Le' dd/MM 'à' HH:mm");

            holder.date.setText(dateFormat_oneday.format(event.getDateStart()));
        }
        else {
            DateFormat dateFormat = new SimpleDateFormat("dd/MM");
            String dateStart = dateFormat.format(event.getDateStart());
            String dateEnd = dateFormat.format(event.getDateEnd());

            holder.date.setText("Du " + dateStart + " au " + dateEnd);
        }

        holder.bind(event, listener);
    }

    @Override
    public int getItemCount() {
        return events.size();
    }

    public List<Event> getEvents() {
        return events;
    }

    public static class EventViewHolder extends RecyclerView.ViewHolder {
        public ImageView thumbnail;
        public TextView name;
        public CircleImageView avatar;
        public TextView date;
        public TextView participants;

        public EventViewHolder(View view) {
            super(view);

            this.avatar = (CircleImageView) view.findViewById(R.id.avatar_club_event);

            this.name = (TextView) view.findViewById(R.id.name_event);
            this.date = (TextView) view.findViewById(R.id.date_event);
            this.participants = (TextView) view.findViewById(R.id.going_event);
            this.thumbnail = (ImageView) view.findViewById(R.id.thumbnail_event);
        }

        public void bind(final Event event, final OnEventItemClickListener listener) {
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onEventItemClick(event);
                }
            });
        }
    }
}

File: app/src/main/java/fr/insapp/insapp/adapters/EventRecyclerViewAdapter.kt
package fr.insapp.insapp.adapters

import android.content.Intent
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.RequestManager
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade
import com.bumptech.glide.request.RequestOptions
import fr.insapp.insapp.R
import fr.insapp.insapp.activities.AssociationActivity
import fr.insapp.insapp.activities.EventActivity
import fr.insapp.insapp.http.ServiceGenerator
import fr.insapp.insapp.models.Association
import fr.insapp.insapp.models.Event
import fr.insapp.insapp.utility.inflate
import kotlinx.android.synthetic.main.row_event.view.*
import kotlinx.android.synthetic.main.row_event_with_avatars.view.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by thomas on 18/11/2016.
 * Kotlin rewrite on 28/08/2019.
 */

class EventRecyclerViewAdapter(
        var events: MutableList<Event>,
        private val requestManager: RequestManager,
        private val past: Boolean,
        private val layout: Int
) : RecyclerView.Adapter<EventRecyclerViewAdapter.EventViewHolder>() {

    fun addItem(event: Event) {
        this.events.add(event)
        Collections.sort(events, EventComparator(past))

        this.notifyDataSetChanged()
    }

    fun updateEvent(position: Int, event: Event) {
        events[position] = event
        notifyItemChanged(position)
    }

    fun removeItem(eventId: String) {
        for (event in events) {
            if (event.id == eventId) {
                events.remove(event)

                this.notifyDataSetChanged()
                return
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = EventViewHolder(parent.inflate(layout), requestManager, layout)

    override fun onBindViewHolder(holder: EventViewHolder, position: Int) {
        val event = events[position]
        holder.bindEvent(event)
    }

    override fun getItemCount() = events.size

    class EventViewHolder(
            private val view: View,
            private val requestManager: RequestManager,
            private val layoutId: Int
    ) : RecyclerView.ViewHolder(view), View.OnClickListener {

        private var event: Event? = null

        init {
            view.setOnClickListener(this)
        }

        fun bindEvent(event: Event) {
            this.event = event

            val context = itemView.context

            if (layoutId == R.layout.row_event_with_avatars) {
                val call = ServiceGenerator.client.getAssociationFromId(event.association)
                call.enqueue(object : Callback<Association> {
                    override fun onResponse(call: Call<Association>, response: Response<Association>) {
                        if (response.isSuccessful) {
                            val association = response.body()

                            if (association != null) {
                                requestManager
                                    .load(ServiceGenerator.CDN_URL + association.profilePicture)
                                    .apply(RequestOptions.circleCropTransform())
                                    .transition(withCrossFade())
                                    .into(view.event_association_avatar)

                                view.event_association_avatar.setOnClickListener { context.startActivity(Intent(context, AssociationActivity::class.java).putExtra("association", association)) }
                            }
                        } else {
                            Toast.makeText(context, "EventRecyclerViewAdapter", Toast.LENGTH_LONG).show()
                        }
                    }

                    override fun onFailure(call: Call<Association>, t: Throwable) {
                        Toast.makeText(context, "EventRecyclerViewAdapter", Toast.LENGTH_LONG).show()
                    }
                })
            }

            requestManager
                .load(ServiceGenerator.CDN_URL + event.image)
                .apply(RequestOptions().transforms(CenterCrop(), RoundedCorners(8)))
                .into(view.event_thumbnail)

            view.event_name.text = event.name

            val nbAttendees = if (event.attendees == null) 0 else event.attendees.size
            if (nbAttendees <= 1) {
                view.event_going.text = String.format(context.getString(R.string.x_attendee), nbAttendees)
            } else {
                view.event_going.text = String.format(context.getString(R.string.x_attendees), nbAttendees)
            }

            val diffInDays = ((event.dateEnd.time - event.dateStart.time) / (1000 * 60 * 60 * 24)).toInt()
            if (diffInDays < 1 && event.dateStart.month == event.dateEnd.month) {
                val dateFormatOneDay = SimpleDateFormat("'Le' dd/MM 'à' HH:mm")

                view.event_date.text = dateFormatOneDay.format(event.dateStart)
            } else {
                val dateFormat = SimpleDateFormat("dd/MM")
                val dateStart = dateFormat.format(event.dateStart)
                val dateEnd = dateFormat.format(event.dateEnd)

                view.event_date.text = "Du $dateStart au $dateEnd"
            }
        }

        override fun onClick(v: View) {
            val context = itemView.context
            context.startActivity(Intent(context, EventActivity::class.java).putExtra("event", event))
        }
    }

    class EventComparator(private val past: Boolean) : Comparator<Event> {

        override fun compare(event1: Event, event2: Event): Int {
            if (event1.dateStart.time == event2.dateStart.time)
                return 0
            if (event1.dateStart.time < event2.dateStart.time)
                return if (past) 1 else -1

            return if (past) -1 else 1
        }
    }
}